java/ql/src/experimental/Security/CWE/CWE-489/ServletMain.ql

```diff
@@ -0,0 +1,51 @@
+/**
+ * @name Main Method in Servlet
+ * @description Jave EE web applications with a main method.
+ * @kind problem
+ * @id java/main-method-in-servlet
+ * @tags security
+ *       external/cwe-489
+ */
+
+import java
+import semmle.code.java.frameworks.Servlets
+
+/** The java type `javax.servlet.Filter` */
+class ServletFilterClass extends Class {
+  ServletFilterClass() { this.getASupertype*().hasQualifiedName("javax.servlet", "Filter") }
+}
+
+/** Listener class in the package `javax.servlet` and `javax.servlet.http` */
+class ServletListenerClass extends Class {
+  // Various listener classes of Java EE such as ServletContextListener. They all have a name ending with the word "Listener".
+  ServletListenerClass() {
+    this.getASupertype*()
+        .getQualifiedName()
+        .regexpMatch([
+            "javax\\.servlet\\.[a-zA-Z]+Listener", "javax\\.servlet\\.http\\.[a-zA-Z]+Listener"
+          ])
+  }
+}
+
+/** The `main` method in `Servlet`. */
+class ServletMainMethod extends Method {
+  ServletMainMethod() {
+    (
+      this.getDeclaringType() instanceof ServletClass or
+      this.getDeclaringType() instanceof ServletFilterClass or
+      this.getDeclaringType() instanceof ServletListenerClass
```

Comment:
You could extend this list to cover more cases, such as  `org.apache.struts.action.Action`, `javax.ejb.EnterpriseBean`.

Reply:
@tamasvajk Thanks for the suggestion. Although EJB became less popular in recent years comparing with other technologies like Spring and microservices in cloud, it's still widely used so I've added a new query for checking the <code>main</code> method in enterprise beans.

Reply:
I've also added query for Struts and Spring actions, which all serve HTTP requests/responses directly.